### PR TITLE
ROS 2 plugins v2 misc fixes

### PR DESCRIPTION
**Pull request by artivis**
This PR is a follow-up on #3536 and essentially fixes the same issues for ROS 2 plugins v2.

This PR most notably,

- changes the colcon install path from `$SNAPCRAFT_PART_INSTALL` to `$SNAPCRAFT_PART_INSTALL/opt/ros/snap` for a cleaner separation of ROS 2 workspaces (similar to what's in the v1 plugin).
- looks for ROS 2 packages in `$SNAPCRAFT_PART_INSTALL/opt/ros/snap` in case the workspace exists (e.g. provided by a stage-snap) during the build step
- does not `rosdep resolve` packages that are available locally in `$SNAPCRAFT_PART_INSTALL` during stage step. Such locally available package may be part of the built sources or provided by a stage-snap. Furthermore, it fixes the following current issue; if the source contains packages `foo` & `bar` - two packages that are not released in the rosdistro database - where `foo` depends on `bar`, staging fails as it tries to `rosdep resolve` the `bar` key (which does not exists since it is not released). With this fix, it finds `bar` locally and thus does not try to resolve it. This issue is fixed by deleting the `${SNAPCRAFT_PART_INSTALL}/opt/ros/snap/COLCON_IGNORE` marker file - if it exists - right after the actual build command. This marker file prevents catkin from crawling the directory while looking for packages during staging.
- adds `test-snapcraft-fake-ros2-package-core20` snap and `colcon-stage-snaps` snap/spread test to test the above bullet points.

Note:

- I own the `test-snapcraft-fake-ros2-package-core20` snap that's used in the new spread test, I imagine we should transfer ownership.

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: tests/spread/plugins/v2/snaps/colcon-stage-snaps/snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+name: colcon-stage-snaps
```
Comment (cjp256) [inline tests/spread/plugins/v2/snaps/colcon-stage-snaps/snap/snapcraft.yaml:1]: Can this be tested with the extension? Looks like the snapcraft.yaml is pre-expanded.  Or is this setup not supported with the extension?

File: tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# What is this, and why isn't it used?
```
Comment (cjp256) [inline tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/README.md:1]: Is the intention that this package is maintained from here going forward? As in, this is the source tree to publish a new test-snapcraft-fake-ros2-package-core20?

File: tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/snap/snapcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+name: test-snapcraft-fake-ros2-package-core20
+base: core20
+version: '0.1'
+summary: Bare-bones ROS 2 dependency for snapcraft CLI test
+description: |
+  This snap contains a single unusable ROS 2 package. This snap is not meant to
```
Comment (cjp256) [inline tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/snap/snapcraft.yaml:6]: Unusable in the sense that it sense that it lacks practical utility?  The app should function, correct?

File: tests/spread/plugins/v2/snaps/colcon-stage-snaps/snap/snapcraft.yaml (hunk 1 above)
Comment (artivis) [reply tests/spread/plugins/v2/snaps/colcon-stage-snaps/snap/snapcraft.yaml:1]: My bad, I was toying with the extension and forgot to revert that. Will fix it.

File: tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/README.md (hunk 2 above)
Comment (artivis) [reply tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/README.md:1]: That's correct; much like it's ROS 1 counterpart [test-snapcraft-fake-ros-package-core20](https://github.com/snapcore/snapcraft/blob/5060014453a8c549aaa79de92a61e28efd0bf14d/tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros-package-core20/README.md).

File: tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/snap/snapcraft.yaml (hunk 3 above)
Comment (artivis) [reply tests/spread/plugins/v2/snaps/test-snapcraft-fake-ros2-package-core20/snap/snapcraft.yaml:6]: That's right, the app function properly but in a way that's only useful to the test.

**Comment by sergiusens** [comment]
@artivis yes it can, but I merged your already passing PR, so it would re-trigger once conflicts are resolved.

**Comment by artivis** [comment]
Merged master and fixed conflicts; tests passed fine :+1:.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**